From 813460ef86e4f81a128ef2b0d83eea8581dc4018 Mon Sep 17 00:00:00 2001
From: Heinrich Toews <heinrich.toews@wago.com>
Date: Wed, 28 Nov 2018 11:40:16 +0000
Subject: [PATCH] net: ethernet: ti: davinci_emac: fix broken mac_addr code
 path

The current code path falsely overwrites the oftree based mac addr
configuration which should always have the highest priority.

This commit changes the behavior to the following:

check if oftree/pdata mac addr is valid
   |
   | if not, try to get it from HW
        |
        | if it fails, get a random addr

Signed-off-by: Heinrich Toews <heinrich.toews@wago.com>
---
 drivers/net/ethernet/ti/davinci_emac.c | 24 ++++++++++++++++--------
 1 file changed, 16 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/ti/davinci_emac.c b/drivers/net/ethernet/ti/davinci_emac.c
index 413cf14..ae487ca 100644
--- a/drivers/net/ethernet/ti/davinci_emac.c
+++ b/drivers/net/ethernet/ti/davinci_emac.c
@@ -1894,16 +1894,24 @@ static int davinci_emac_probe(struct platform_device *pdev)
 	}
 	ndev->irq = res->start;
 
-	rc = davinci_emac_try_get_mac(pdev, res_ctrl ? 0 : 1, priv->mac_addr);
-	if (!rc)
-		ether_addr_copy(ndev->dev_addr, priv->mac_addr);
 
+	/* if no mac addr was passed first try to get it from HW,
+	 * if it fails get a random addr
+	 */
 	if (!is_valid_ether_addr(priv->mac_addr)) {
-		/* Use random MAC if none passed */
-		eth_hw_addr_random(ndev);
-		memcpy(priv->mac_addr, ndev->dev_addr, ndev->addr_len);
-		dev_warn(&pdev->dev, "using random MAC addr: %pM\n",
-							priv->mac_addr);
+		rc = davinci_emac_try_get_mac(pdev, res_ctrl ? 0 : 1,
+					      priv->mac_addr);
+		if (rc || !is_valid_ether_addr(priv->mac_addr)) {
+			/* Use random MAC if none passed */
+			eth_hw_addr_random(ndev);
+			memcpy(priv->mac_addr, ndev->dev_addr, ndev->addr_len);
+			dev_warn(&pdev->dev,
+				 "using random MAC addr: %pM\n",
+				 priv->mac_addr);
+		} else {
+			/* set MAC address from HW */
+			ether_addr_copy(ndev->dev_addr, priv->mac_addr);
+		}
 	}
 
 	ndev->netdev_ops = &emac_netdev_ops;
-- 
2.7.4

